### DISCUSSION TITLE
LUA for Beginners: How to send controller values to midi device?

### DISCUSSION THREAD

**Discussion by nemesis2049** [RomanKubiak/ctrlr#578] [Q&A] [answered]
Hello everyone, I have created a panel for the Kawai SX-240 and would like to have a way to send the current controller data to the synth. They are all CC messages. Snapshot is doesnt work.

I've been trying many things for two days now, but I'm not really making any progress as to whether I can program (PHP, JS, C++). What's the easiest way to send a current modulator value to the synth with a CC-MIDI message?

Is there somewhere a good overview of all the methods and attributes I can use in CTRLR?

Thanks so far!

**Comment by dobo365** [comment]
Hi,
  not sure I fully understand your current situation..
  You created some modulators but they are not sending CC messages to your synth?
  Or you want to be able to send all parameters at once to your synth?
  
  Please have a look to my Ctrlr Step by Step guide even if a bit "old". I thought it was still accessible on ctrlr.org but apparently not as you must be logged or have the link.
  So, I attach it here (will create a Ctrlr section on my sunnysynths.com web site to make it all the time accessible.
[Ctrlr - Step by Step - v13.pdf](https://github.com/RomanKubiak/ctrlr/files/13347465/Ctrlr.-.Step.by.Step.-.v13.pdf)

I'm explaining how to send CC.

If you want to send all parameters at once, you will need to write a Lua method taking each modulators value and sending it to the synth.

Good luck!

**Comment by dnaldoog** [comment, marked as answer]
Hi @nemesis2049 

- **Edit**: Or as Goodweather said, not sure if you need this or just need to send CC? Anyway - you can see how to do that in this panel.

Here is one suggestion for how to do this.


Substitute your modulator names for the ones I have in `snapshotTable`  and it should work for you. 

<pre>
--
-- Called when a mouse is down on this component
--

myMethod = function(comp,event)
    local delay = panel:getModulatorByName("sendRate"):getValueMapped() -- milliseconds (adjust to suit)
    ch = panel:getProperty("panelMidiOutputChannelDevice") - 1 -- global
    counter = 0 -- global
    timer:setCallback(1, TimerCallback) -- TimerId 1
    timer:startTimer(1, delay) -- Timer delay in milliseconds
end

snapshotTable = { -- add your modulator names here - that's all :)
    "LFO rate",
    "LFO delay",
    "attack",
    "decay",
    "sustain",
    "release"
}

TimerCallback = function(timerId)
    if timerId == 1 then
        counter = counter + 1 -- keep global
        local num = panel:getModulatorByName(snapshotTable[counter]):getMidiMessage():getProperty("midiMessageCtrlrNumber")
        local value = panel:getModulatorByName(snapshotTable[counter]):getModulatorValue()
        local sysm = string.format("%.2x %.2x %.2x", 176 + ch, num, value)
        panel:sendMidiMessageNow(CtrlrMidiMessage(sysm))
    end
    if counter == #snapshotTable then
        timer:stopTimer(timerId)
    end
end 

</pre>
----
[Send Snapshot CC_1_0.zip](https://github.com/RomanKubiak/ctrlr/files/13348015/Send.Snapshot.CC_1_0.zip)

**Reply by nemesis2049** [reply]
Hi guys, thanks for your support! 

I'm looking for a LUA-script which is sending all modulator-parameters (the patch) to the synth once. I need a possibility to send all values back to the synth after opening the track in my DAW.

Usually the snapshout-function should do it, but it doesnt work with my synth. So i decide to program a "Send Button" which is sending all modulator values to the synth.

Something like this:

```
function getModulatorValue(modulatorID)
    local modulator = panel:getModulatorByIndex(modulatorID)
    if modulator ~= nil then
        return modulator:getValue()
    else
        return nil
    end
end

-- Example for sending cutoff value (CC74, Mod-ID 1)
midiMessage = {
   channel = 1,
   type = CONTROL_CHANGE,
   controllerNumber = 74, 
   value = getModulatorValue(1)
}
panel:sendMessage (midiMessage)
```

But this drops an error. Using this `panel:sendMidiMessageNow(CtrlrMidiMessage(midiMessage))` brings no error, but doesnt work too. Ive no idea what the right synthax is for sending a MIDI-message. Is `panel` a global value? Can i use it in the script?

So much questions ... 

Thanks

**Reply by nemesis2049** [reply]
@dnaldoog 

Your script looks interesting and is exactly what I'am looking for.
I will check it with my panel + synth!

**Reply by nemesis2049** [reply]
@dnaldoog 

Your solution seems to work. Thanks!!
If i import the method into my own panel it crashes after sending the button. 
So, there is some issue with panel or its config! 

I will keep you updated!

**Reply by dnaldoog** [reply]
@nemesis2049 Make sure you are using either version 5.3.201 for Windows or 5.3.198 on macOS

**Reply by nemesis2049** [reply]
I'm using the latest version 6.04 and it works. (VST3)
Clicking on the snapshot-library in the menu crashes, but i dont need it.

**Reply by dobo365** [reply]
I think I have read somewhere that snapshots were not supported anymore.
Anyway, never used that.

**Comment by dnaldoog** [comment]
This version of your code would work. It's a bit unusual for me but interesting!

<pre>
myMethod = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
    local x = getModulatorValue(1)
    local ch = panel:getProperty("panelMidiOutputChannelDevice") - 1 
    -- Example for sending cutoff value (CC74, Mod-ID 1)
   local  midiMessage = {
        0xB0 + ch, -- channel = 1,
        74, -- controllerNumber = 74,
        x --value = getModulatorValue(1)
    }
    panel:sendMidiMessageNow(CtrlrMidiMessage(midiMessage))
end

function getModulatorValue(modulatorID)
    local modulator = panel:getModulatorByIndex(modulatorID)
    if modulator ~= nil then
        return modulator:getValue()
    else
        return nil
    end
end

</pre>

**Reply by nemesis2049** [reply]
Sorry, iam new to LUA and with my first steps i try to understand the classes and methods.
But anyway, your code worked fine for my project. Now i can use it inside my DAW projects. This is great!

Next days i will dive deeper into it and optimize some things in the editor. 
If anyone is interested in the SX-240 editor - let me know it.

@dobo365 
Really good guide!!

Thanks so much for your support!

**Comment by dobo365** [comment]
As usual, thx also to John with his detailed explanations (I'm sorry to not have that time).

One thing to pay attention in your method is that **only the dynamic modulators have a Midi section** (labels do not have a Midi section) and your code should check that to avoid throwing an error.
So, you need to add a check like this one:

```
if mod:getProperty("modulatorIsStatic") == "0" then --	 Only Dynamic modulators have a MidiMessage ;-)
	-- your code
end
```

I'm using this to Mute/Unmute all dynamic modulators when receiving a program from the synth which will adapt all dynamic modulators and that I don't want to have the CC messages sent back to the synth.
I'm also lazy on that one and take/scan all modulators by ID.
Of course, using John's technique with a lookup table avoid this issue as you would only put the good modulators in the table.

**Comment by dnaldoog** [comment]
Didn’t know about

if mod:getProperty("modulatorIsStatic") == "0" then --	 Only Dynamic modulators have a MidiMessage ;-)
	-- your code
end

Thank you  Dominique !

**Reply by dobo365** [reply]
I was guessing you were not aware of that one ;-)
My pleasure!

**Comment by dnaldoog** [comment]
It occurred to me that the timer code should probably be:  

<pre>
TimerCallback = function(timerId)
    if timerId == 1 then
        counter = counter + 1 -- keep global
        local num = panel:getModulatorByName(snapshotTable[counter]):getMidiMessage():getProperty("midiMessageCtrlrNumber")
        local value = panel:getModulatorByName(snapshotTable[counter]):getModulatorValue()
        local sysm = string.format("%.2x %.2x %.2x", 176 + ch, num, value)
        panel:sendMidiMessageNow(CtrlrMidiMessage(sysm))
  
    if counter == #snapshotTable then
        timer:stopTimer(timerId)
      end
    end
end 
</pre>
Not sure if the previous code contains a subtle bug.
